rest/authorization-server/src/main/java/org/eclipse/sw360/rest/authserver/security/customheaderauth/Sw360CustomHeaderAuthenticationProvider.java

```diff
@@ -0,0 +1,164 @@
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.AuthenticationException;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+import org.springframework.security.oauth2.common.util.OAuth2Utils;
+import org.springframework.security.oauth2.provider.ClientDetails;
+import org.springframework.security.oauth2.provider.ClientDetailsService;
+import org.springframework.security.oauth2.provider.ClientRegistrationException;
+import org.springframework.security.web.authentication.preauth.PreAuthenticatedAuthenticationToken;
+
+import javax.annotation.PostConstruct;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * This {@link AuthenticationProvider} is specialized on requests where the
+ * {@link Sw360CustomHeaderAuthenticationFilter} made sure that the user has
+ * already been authenticated by some external proxy that set some headers to
+ * let us know about the authentication.
+ *
+ * In addition it is special because it calculates the granted authorities for
+ * the user depending on the user's authorities given by his groups and the
+ * client's scopes. The result will be the intersection between these two lists.
+ * Of course this is only done for an oauth request and not for normal ones
+ * (that have nothing to do with clients).
+ */
+public class Sw360CustomHeaderAuthenticationProvider implements AuthenticationProvider {
+
+    private final Logger log = Logger.getLogger(this.getClass());
+
+    @Value("${security.customheader.headername.intermediateauthstore:#{null}}")
+    private String customHeaderHeadernameIntermediateAuthStore;
+
+    @Autowired
+    private Sw360UserDetailsProvider sw360CustomHeaderUserDetailsProvider;
+
+    @Autowired
+    private ClientDetailsService clientDetailsService;
+
+    private boolean active;
+
+    @PostConstruct
+    public void postSw360CustomHeaderAuthenticationProviderConstruction() {
+        if (StringUtils.isEmpty(customHeaderHeadernameIntermediateAuthStore)) {
+            log.warn("AuthenticationProvider is NOT active! Some configuration is missing. Needed config keys:\n"
+                    + "- security.customheader.headername.intermediateauthstore");
+            active = false;
+        } else {
+            log.info("AuthenticationProvider is active!");
+            active = true;
+        }
+    }
+
+    @Override
+    public Authentication authenticate(Authentication authentication) throws AuthenticationException {
+        // check if the marker header of our filter is available
+        if (authentication.getDetails() instanceof Map<?, ?>
+                && ((Map<?, ?>) authentication.getDetails()).containsKey(customHeaderHeadernameIntermediateAuthStore)) {
+            Map<?, ?> authDetails = ((Map<?, ?>) authentication.getDetails());
+
+            // get user details
```

Comment:
Instead of comment, maybe extract into function with name `getUserDetails`

Reply:
Some of the code extracted into own class, since other code duplicated it as well: see `Sw360UserAndClientAuthoritiesMerger`